source/slang/slang-ir-lower-generics.cpp

```diff
@@ -382,110 +437,146 @@ namespace Slang
-                                rawPtrType,
-                                arg);
-                        }
-                        args.add(arg);
-                    }
-                    for (UInt i = 0; i < specializeInst->getArgCount(); i++)
-                    {
-                        auto arg = specializeInst->getArg(i);
-                        // Translate Type arguments into RTTI object.
-                        if (as<IRType>(arg))
-                        {
-                            // We are using a simple type to specialize a callee.
-                            // Generate RTTI for this type.
-                            auto rttiObject = maybeEmitRTTIObject(arg);
-                            arg = builder->emitGetAddress(
-                                builder->getPtrType(builder->getRTTIType()),
-                                rttiObject);
-                        }
-                        else if (arg->op == kIROp_Specialize)
-                        {
-                            // The type argument used to specialize a callee is itself a
-                            // specialization of some generic type.
-                            // TODO: generate RTTI object for specializations of generic types.
-                            SLANG_UNIMPLEMENTED_X("RTTI object generation for generic types");
-                        }
-                        else if (arg->op == kIROp_RTTIObject)
-                        {
-                            // We are inside a generic function and using a generic parameter
-                            // to specialize another callee. The generic parameter of the caller
-                            // has already been translated into an RTTI object, so we just need
-                            // to pass this object down.
-                        }
-                        args.add(arg);
-                    }
-                    auto newCall = builder->emitCallInst(callInst->getFullType(), loweredFunc, args);
-                    callInst->replaceUsesWith(newCall);
-                    callInst->removeAndDeallocate();
+            for (UInt i = 0; i < funcType->getParamCount(); i++)
+                paramTypes.add(funcType->getParamType(i));
+
+            IRBuilder builderStorage;
+            auto builder = &builderStorage;
+            builder->sharedBuilder = &sharedBuilderStorage;
+            builder->setInsertBefore(callInst);
+
+            List<IRInst*> args;
+
+            // Indicates whether the caller should allocate space for return value.
+            // If the lowered callee returns void and this call inst has a type that is not void,
+            // then we are calling a transformed function that expects caller allocated return value
+            // as the first argument.
+            bool shouldCallerAllocateReturnValue = (funcType->getResultType()->op == kIROp_VoidType &&
+                callInst->getDataType() != funcType->getResultType());
```

Comment:
This is more of the logic that could be pulled out into its own pass instead, right?

Reply:
You are right. I may do it together with my associated types change.